### PR TITLE
Add support for querying specific namespaces

### PR DESCRIPTION
**Pull request by shreyassrivatsan**
**What this PR does / why we need it**:
This change adds support for querying specific namespaces by
means of headers. It additionally adds support for graphite to use
the full fetch options.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:
```documentation-note
NONE
```

### CODE REVIEW

**Comment by linasm** [comment]
Wouldn't it be possible to adapt `MetricsStoragePolicyHeader` to support those new formats (in addition to the currently supported `xx:yy` format? I think it would be less confusing to have one header instead of two functionally overlapping.

**Comment by linasm** [comment]
Nvm, `MetricsStoragePolicyHeader` supporting writes in addition to reads makes my point somewhat moot. Feel free to ignore it.

File: src/query/storage/restrict_query_options.go [hunk 1]
```diff
@@ -34,7 +34,13 @@ func (o *RestrictQueryOptions) Validate() error {
 	if o.RestrictByType != nil {
 		return o.RestrictByType.Validate()
 	}
-
+	if len(o.RestrictByTypes) > 0 {
```
Comment (ryanhall07) [inline src/query/storage/restrict_query_options.go:37]: nit: don't need the if check

File: src/query/storage/types.go [hunk 2]
```diff
@@ -205,6 +205,9 @@ type RestrictQueryOptions struct {
 	// RestrictByTag are specific restrictions to enforce behavior for given
 	// tags.
 	RestrictByTag *RestrictByTag
+	// RestrictByTypes are specific restrictions to stick to different data
```
Comment (ryanhall07) [inline src/query/storage/types.go:208]: i don't follow what "to stick to" means

File: src/x/headers/headers.go [hunk 3]
```diff
@@ -76,6 +76,11 @@ const (
 	// metrics type.
 	MetricsStoragePolicyHeader = M3HeaderPrefix + "Storage-Policy"
 
+	// MetricsRestrictByPoliciesHeader provides the policies options to enforce
+	// on queries, in the form of a list of storage policies.
+	// "1m:14d;5m:60d"
+	MetricsRestrictByPoliciesHeader = M3HeaderPrefix + "Restrict-By-Policies"
```
Comment (ryanhall07) [inline src/x/headers/headers.go:82]: why `Policies` and not `Namespaces`?

Comment (ryanhall07) [reply src/x/headers/headers.go:82]: ah it's a storage policy? can we make the name more explicit to match `M3-Storage-Policy`

**Review by ryanhall07 — APPROVED** [review]
a couple of comments

File: src/x/headers/headers.go (hunk 3 above)
Comment (shreyassrivatsan) [reply src/x/headers/headers.go:82]: Looks like we use storage policies as our general terminology everywhere, so went with policies.

Comment (ryanhall07) [reply src/x/headers/headers.go:82]: that's fine. but I think we should be more explicit and use `M3-Restrict-By-Storage-Policies`

Comment (shreyassrivatsan) [reply src/x/headers/headers.go:82]: Ah yes, missed this. renaming [here](https://github.com/m3db/m3/pull/3719).

If both ways are set we fail the query.

**Pull request merged**